src/ChangeStream.php

```diff
@@ -0,0 +1,106 @@
+<?php
+/*
+ * Copyright 2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB;
+
+use MongoDB\Driver\Cursor;
+use MongoDB\Driver\Exception\RuntimeException;
+use MongoDB\Exception\ResumeTokenException;
+use IteratorIterator;
+use Iterator;
+
+/**
+ * Iterator for the changeStream command.
+ *
+ * @api
+ * @see \MongoDB\Collection::changeStream()
+ * @see http://docs.mongodb.org/manual/reference/command/changeStream/
+ */
+class ChangeStream implements Iterator
+{
+    private $resumeToken;
+    private $resumeCallable;
+    private $csIt;
+    private $key;
+
+    public function __construct(Cursor $cursor, callable $resumeCallable)
+    {
+        $this->resumeCallable = $resumeCallable;
+        $this->csIt = new IteratorIterator($cursor);
+        $this->csIt->rewind();
+        $this->key = 0;
+    }
+
+    public function current()
+    {
+        return $this->csIt->current();
+    }
+
+    public function getId()
+    {
+        return $this->csIt->getInnerIterator()->getId();
+    }
+
+    public function key()
+    {
+        return $this->key;
+    }
+
+    public function next()
+    {
+        $this->key++;
```

Comment:
Not every advancement with `next()` is going to be successful. This looks like it's possible for `$this->key` to get incremented more times than necessary.

I think a test where there is no next element (because no changes have occurred) or we end up resuming a few times would demonstrate that the key can get out of sync.